command/agent/http.go

```diff
@@ -75,89 +76,111 @@ type HTTPServer struct {
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
 
-	wsUpgrader := &websocket.Upgrader{
-		ReadBufferSize:  2048,
-		WriteBufferSize: 2048,
-	}
+			tlsConfig, err := tlsConf.IncomingTLSConfig()
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
+			ln = tls.NewListener(tcpKeepAliveListener{ln.(*net.TCPListener)}, tlsConfig)
+		}
 
-	// Create the server
-	srv := &HTTPServer{
-		agent:      agent,
-		mux:        mux,
-		listener:   ln,
-		listenerCh: make(chan struct{}),
-		logger:     agent.httpLogger,
-		Addr:       ln.Addr().String(),
-		wsUpgrader: wsUpgrader,
-	}
-	srv.registerHandlers(config.EnableDebug)
+		// Create the mux
+		mux := http.NewServeMux()
 
-	// Handle requests with gzip compression
-	gzip, err := gziphandler.GzipHandlerWithOpts(gziphandler.MinSize(0))
-	if err != nil {
-		return nil, err
-	}
+		wsUpgrader := &websocket.Upgrader{
+			ReadBufferSize:  2048,
+			WriteBufferSize: 2048,
+		}
 
-	// Get connection handshake timeout limit
-	handshakeTimeout, err := time.ParseDuration(config.Limits.HTTPSHandshakeTimeout)
-	if err != nil {
-		return nil, fmt.Errorf("error parsing https_handshake_timeout: %v", err)
-	} else if handshakeTimeout < 0 {
-		return nil, fmt.Errorf("https_handshake_timeout must be >= 0")
-	}
+		// Create the server
+		srv := &HTTPServer{
+			agent:      agent,
+			mux:        mux,
+			listener:   ln,
+			listenerCh: make(chan struct{}),
+			logger:     agent.httpLogger,
+			Addr:       ln.Addr().String(),
+			wsUpgrader: wsUpgrader,
+		}
+		srv.registerHandlers(config.EnableDebug)
 
-	// Get max connection limit
-	maxConns := 0
-	if mc := config.Limits.HTTPMaxConnsPerClient; mc != nil {
-		maxConns = *mc
-	}
-	if maxConns < 0 {
-		return nil, fmt.Errorf("http_max_conns_per_client must be >= 0")
-	}
+		// Handle requests with gzip compression
+		gzip, err := gziphandler.GzipHandlerWithOpts(gziphandler.MinSize(0))
+		if err != nil {
+			error = multierror.Append(error, err)
+			continue
+		}
+
+		// Get connection handshake timeout limit
+		handshakeTimeout, err := time.ParseDuration(config.Limits.HTTPSHandshakeTimeout)
+		if err != nil {
+			error = multierror.Append(error, fmt.Errorf("error parsing https_handshake_timeout: %v", err))
+			continue
+		} else if handshakeTimeout < 0 {
+			error = multierror.Append(error, fmt.Errorf("https_handshake_timeout must be >= 0"))
+			continue
+		}
 
-	// Create HTTP server with timeouts
-	httpServer := http.Server{
-		Addr:      srv.Addr,
-		Handler:   gzip(mux),
-		ConnState: makeConnState(config.TLSConfig.EnableHTTP, handshakeTimeout, maxConns),
-		ErrorLog:  newHTTPServerLogger(srv.logger),
+		// Get max connection limit
+		maxConns := 0
+		if mc := config.Limits.HTTPMaxConnsPerClient; mc != nil {
+			maxConns = *mc
+		}
+		if maxConns < 0 {
+			error = multierror.Append(error, fmt.Errorf("http_max_conns_per_client must be >= 0"))
+			continue
+		}
```

Comment:
I think it's safe to put this whole section outside the loop so that we don't have to mess with multierrors and can return early if we're misconfigured.